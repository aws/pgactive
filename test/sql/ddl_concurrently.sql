\c regression

SELECT pgactive.pgactive_replicate_ddl_command($DDL$
CREATE TABLE public.concurrently_test (
	id integer not null primary key
);
$DDL$);

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
\c postgres

\d public.concurrently_test

-- Fails: ddl rep not skipped
DROP INDEX CONCURRENTLY concurrently_test_pkey;

-- Fails: ddl rep not skipped
CREATE INDEX CONCURRENTLY named_index ON concurrently_test(id);

-- Fails: drop the constraint
SET pgactive.skip_ddl_replication = on;
DROP INDEX CONCURRENTLY concurrently_test_pkey;
RESET pgactive.skip_ddl_replication;

-- Fails: no direct DDL
ALTER TABLE public.concurrently_test
DROP CONSTRAINT concurrently_test_pkey;

-- succeeds
SELECT pgactive.pgactive_replicate_ddl_command($DDL$
ALTER TABLE public.concurrently_test
DROP CONSTRAINT concurrently_test_pkey;
$DDL$);

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

-- We can create a new index
SET pgactive.skip_ddl_replication = on;
CREATE INDEX CONCURRENTLY named_index ON concurrently_test(id);
RESET pgactive.skip_ddl_replication;

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);

\c regression

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

SET pgactive.skip_ddl_replication = on;
CREATE INDEX CONCURRENTLY named_index ON concurrently_test(id);
RESET pgactive.skip_ddl_replication;

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);

\c postgres

-- Fails, no skip ddl rep
DROP INDEX CONCURRENTLY named_index;

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

-- ok
SET pgactive.skip_ddl_replication = on;
DROP INDEX CONCURRENTLY named_index;
RESET pgactive.skip_ddl_replication;

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);

\c regression

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;

-- Have to drop on each node
SET pgactive.skip_ddl_replication = on;
DROP INDEX CONCURRENTLY named_index;
RESET pgactive.skip_ddl_replication;

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
