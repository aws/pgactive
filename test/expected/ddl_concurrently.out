\c regression
SELECT pgactive.pgactive_replicate_ddl_command($DDL$
CREATE TABLE public.concurrently_test (
	id integer not null primary key
);
$DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
-----------------------
 
 
(2 rows)

\c postgres
\d public.concurrently_test
Table "public.concurrently_test"
 Column |  Type   | Modifiers 
--------+---------+-----------
 id     | integer | not null
Indexes:
    "concurrently_test_pkey" PRIMARY KEY, btree (id)

-- Fails: ddl rep not skipped
DROP INDEX CONCURRENTLY concurrently_test_pkey;
ERROR:  DROP INDEX CONCURRENTLY is not supported without pgactive.skip_ddl_replication set
-- Fails: ddl rep not skipped
CREATE INDEX CONCURRENTLY named_index ON concurrently_test(id);
ERROR:  CREATE INDEX CONCURRENTLY without pgactive.skip_ddl_replication set may only affect UNLOGGED or TEMPORARY tables when pgactive is active; concurrently_test is a regular table
-- Fails: drop the constraint
SET pgactive.skip_ddl_replication = on;
DROP INDEX CONCURRENTLY concurrently_test_pkey;
ERROR:  cannot drop index concurrently_test_pkey because constraint concurrently_test_pkey on table concurrently_test requires it
HINT:  You can drop constraint concurrently_test_pkey on table concurrently_test instead.
RESET pgactive.skip_ddl_replication;
-- Fails: no direct DDL
ALTER TABLE public.concurrently_test
DROP CONSTRAINT concurrently_test_pkey;
ERROR:  Direct DDL commands are not supported while pgactive is active
HINT:  Use pgactive.pgactive_replicate_ddl_command(...)
-- succeeds
SELECT pgactive.pgactive_replicate_ddl_command($DDL$
ALTER TABLE public.concurrently_test
DROP CONSTRAINT concurrently_test_pkey;
$DDL$);
 pgactive_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
 relname 
---------
(0 rows)

-- We can create a new index
SET pgactive.skip_ddl_replication = on;
CREATE INDEX CONCURRENTLY named_index ON concurrently_test(id);
RESET pgactive.skip_ddl_replication;
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
   relname   
-------------
 named_index
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
-----------------------
 
 
(2 rows)

\c regression
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
 relname 
---------
(0 rows)

SET pgactive.skip_ddl_replication = on;
CREATE INDEX CONCURRENTLY named_index ON concurrently_test(id);
RESET pgactive.skip_ddl_replication;
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
   relname   
-------------
 named_index
(1 row)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
-----------------------
 
 
(2 rows)

\c postgres
-- Fails, no skip ddl rep
DROP INDEX CONCURRENTLY named_index;
ERROR:  DROP INDEX CONCURRENTLY is not supported without pgactive.skip_ddl_replication set
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
   relname   
-------------
 named_index
(1 row)

-- ok
SET pgactive.skip_ddl_replication = on;
DROP INDEX CONCURRENTLY named_index;
RESET pgactive.skip_ddl_replication;
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
 relname 
---------
(0 rows)

SELECT pgactive.pgactive_wait_for_slots_confirmed_flush_lsn(NULL,NULL);
 pgactive_wait_for_slots_confirmed_flush_lsn 
-----------------------
 
 
(2 rows)

\c regression
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
   relname   
-------------
 named_index
(1 row)

-- Have to drop on each node
SET pgactive.skip_ddl_replication = on;
DROP INDEX CONCURRENTLY named_index;
RESET pgactive.skip_ddl_replication;
SELECT relname FROM pg_class WHERE relname IN ('named_index', 'concurrently_test_pkey') AND relkind = 'i' ORDER BY relname;
 relname 
---------
(0 rows)

